packages/pigeon/lib/src/dart/dart_generator.dart

```diff
@@ -79,28 +85,340 @@ class DartOptions {
+        nullable: type.isNullable,
+        varName: name,
+        code:
+            '$name${_getForceNonNullSymbol(type.isNullable && forceNonNull)}.toJni()',
+      );
+    } else if (!type.isNullable &&
+        (type.baseName == 'int' ||
+            type.baseName == 'double' ||
+            type.baseName == 'bool')) {
+      return name;
+    }
+    return '_PigeonJniCodec.writeValue<${getJniCallReturnType(true)}>($name)';
+  }
+
+  String getJniCallReturnType(bool forceUnwrap) {
+    if (forceUnwrap || type.isNullable || nonNullableNeedsUnwrapping) {
+      return '$jniName$jniCollectionTypeAnnotations${_getNullableSymbol(type.isNullable)}';
+    }
+    return type.baseName;
+  }
+
+  String getDartReturnType(bool forceUnwrap) {
+    if (forceUnwrap || type.isNullable || nonNullableNeedsUnwrapping) {
+      return '${type.baseName}$dartCollectionTypeAnnotations${_getNullableSymbol(type.isNullable)}';
+    }
+    return type.baseName;
+  }
+
+  String get dartCollectionTypeAnnotations {
+    if (type.baseName == 'List') {
+      return '<$dartCollectionTypes>';
+    } else if (type.baseName == 'Map') {
+      return '<$dartCollectionTypes>';
+    }
+    return '';
+  }
+
+  String get jniCollectionTypeAnnotations {
+    if (type.baseName == 'List') {
+      return '<$jniCollectionTypes>';
+    } else if (type.baseName == 'Map') {
+      return '<$jniCollectionTypes>';
+    }
+    return '';
+  }
+
+  String get dartCollectionTypes {
+    if (type.baseName == 'List') {
+      return subTypeOne?.getDartReturnType(true) ?? 'Object?';
+    } else if (type.baseName == 'Map') {
+      return '${subTypeOne?.getDartReturnType(true) ?? 'Object?'}, ${subTypeTwo?.getDartReturnType(true) ?? 'Object?'}';
+    }
+    return '';
+  }
+
+  String get jniCollectionTypes {
+    if (type.baseName == 'List') {
+      return subTypeOne?.getJniCallReturnType(true) ?? 'JObject?';
+    } else if (type.baseName == 'Map') {
+      return '${subTypeOne?.getJniCallReturnType(true) ?? 'JObject'}, ${subTypeTwo?.getJniCallReturnType(true) ?? 'JObject?'}';
+    }
+    return '';
+  }
+}
+
+String _jniArgumentsToDartArguments(
+  List<Parameter> parameters, {
+  bool isAsynchronous = false,
+}) {
+  final List<String> argumentSignature = <String>[];
+  for (final Parameter parameter in parameters) {
+    final _JniType jniType = _JniType.fromTypeDeclaration(parameter.type);
+    argumentSignature.add(
+      jniType.getToDartCall(
+        parameter.type,
+        varName: parameter.name,
+        forceConversion: isAsynchronous,
+      ),
+    );
+  }
+  return argumentSignature.join(', ');
 }
 
+String _getNullableSymbol(bool nullable) => nullable ? '?' : '';
+
+String _getForceNonNullSymbol(bool force) => force ? '!' : '';
+
+String _wrapInNullCheckIfNullable({
+  required bool nullable,
+  required String varName,
+  required String code,
+  String ifNull = 'null',
+}) => nullable ? '$varName == null ? $ifNull : $code' : code;
```

Comment:
Shouldn't the expression form of this be wrapped in `()` so it will work anywhere?